File: packages/elixir/lib/html_to_markdown/options.ex
defmodule HtmlToMarkdown.Options do
  @moduledoc """
  High-level configuration for HTML → Markdown conversion.

  Use `new/1` to build options from a map or keyword list, or construct
  the struct directly. Pass the struct to `HtmlToMarkdown.convert/2`,
  `HtmlToMarkdown.options/1`, or `HtmlToMarkdown.convert_with_inline_images/3`.
  """

  alias HtmlToMarkdown.PreprocessingOptions

  @heading_styles [:underlined, :atx, :atx_closed]
  @list_indent_types [:spaces, :tabs]
  @highlight_styles [:double_equal, :html, :bold, :none]
  @whitespace_modes [:normalized, :strict]
  @newline_styles [:spaces, :backslash]
  @code_block_styles [:indented, :backticks, :tildes]

  @option_keys [
    :heading_style,
    :list_indent_type,
    :list_indent_width,
    :bullets,
    :strong_em_symbol,
    :escape_asterisks,
    :escape_underscores,
    :escape_misc,
    :escape_ascii,
    :code_language,
    :encoding,
    :autolinks,
    :default_title,
    :keep_inline_images_in,
    :br_in_tables,
    :hocr_spatial_tables,
    :highlight_style,
    :extract_metadata,
    :whitespace_mode,
    :strip_newlines,
    :wrap,
    :wrap_width,
    :strip_tags,
    :preserve_tags,
    :convert_as_inline,
    :sub_symbol,
    :sup_symbol,
    :newline_style,
    :code_block_style,
    :preprocessing,
    :debug,
    :skip_images
  ]

  defstruct heading_style: :atx,
            list_indent_type: :spaces,
            list_indent_width: 2,
            bullets: "-*+",
            strong_em_symbol: "*",
            escape_asterisks: false,
            escape_underscores: false,
            escape_misc: false,
            escape_ascii: false,
            code_language: "",
            encoding: "utf-8",
            autolinks: true,
            default_title: false,
            keep_inline_images_in: MapSet.new(),
            br_in_tables: false,
            hocr_spatial_tables: true,
            highlight_style: :double_equal,
            extract_metadata: true,
            whitespace_mode: :normalized,
            strip_newlines: false,
            wrap: false,
            wrap_width: 80,
            strip_tags: MapSet.new(),
            preserve_tags: MapSet.new(),
            convert_as_inline: false,
            sub_symbol: "",
            sup_symbol: "",
            newline_style: :spaces,
            code_block_style: :backticks,
            preprocessing: %PreprocessingOptions{},
            debug: false,
            skip_images: false

  @type heading_style :: :underlined | :atx | :atx_closed
  @type list_indent_type :: :spaces | :tabs
  @type highlight_style :: :double_equal | :html | :bold | :none
  @type whitespace_mode :: :normalized | :strict
  @type newline_style :: :spaces | :backslash
  @type code_block_style :: :indented | :backticks | :tildes

  @type t :: %__MODULE__{
          heading_style: heading_style(),
          list_indent_type: list_indent_type(),
          list_indent_width: pos_integer(),
          bullets: String.t(),
          strong_em_symbol: String.t(),
          escape_asterisks: boolean(),
          escape_underscores: boolean(),
          escape_misc: boolean(),
          escape_ascii: boolean(),
          code_language: String.t(),
          encoding: String.t(),
          autolinks: boolean(),
          default_title: boolean(),
          keep_inline_images_in: MapSet.t(),
          br_in_tables: boolean(),
          hocr_spatial_tables: boolean(),
          highlight_style: highlight_style(),
          extract_metadata: boolean(),
          whitespace_mode: whitespace_mode(),
          strip_newlines: boolean(),
          wrap: boolean(),
          wrap_width: pos_integer(),
          strip_tags: MapSet.t(),
          preserve_tags: MapSet.t(),
          convert_as_inline: boolean(),
          sub_symbol: String.t(),
          sup_symbol: String.t(),
          newline_style: newline_style(),
          code_block_style: code_block_style(),
          preprocessing: PreprocessingOptions.t(),
          debug: boolean(),
          skip_images: boolean()
        }

  @doc """
  Build options from a struct, keyword list, or map.
  """
  @spec new(t() | map() | keyword() | nil) :: t()
  def new(nil), do: %__MODULE__{}
  def new(%__MODULE__{} = opts), do: opts

  def new(attrs) when is_list(attrs) or is_map(attrs) do
    attrs =
      attrs
      |> Map.new()
      |> Enum.reduce(%{}, &normalize_entry/2)

    preprocessing = Map.get(attrs, :preprocessing, nil)
    attrs = Map.delete(attrs, :preprocessing)

    %__MODULE__{}
    |> struct(attrs)
    |> Map.update!(:preprocessing, fn existing ->
      PreprocessingOptions.new(preprocessing || existing)
    end)
  end

  defp normalize_entry({key, value}, acc) do
    case normalize_key(key) do
      nil ->
        acc

      normalized_key ->
        Map.put(acc, normalized_key, normalize_value(normalized_key, value))
    end
  end

  defp normalize_key(key) when is_atom(key), do: if(key in @option_keys, do: key, else: nil)

  defp normalize_key(key) when is_binary(key) do
    key
    |> String.trim()
    |> String.replace("-", "_")
    |> String.to_existing_atom()
  rescue
    _ -> nil
  end

  defp normalize_key(_), do: nil

  defp normalize_value(:heading_style, value), do: normalize_enum(value, @heading_styles, :atx)

  defp normalize_value(:list_indent_type, value),
    do: normalize_enum(value, @list_indent_types, :spaces)

  defp normalize_value(:highlight_style, value),
    do: normalize_enum(value, @highlight_styles, :double_equal)

  defp normalize_value(:whitespace_mode, value),
    do: normalize_enum(value, @whitespace_modes, :normalized)

  defp normalize_value(:newline_style, value), do: normalize_enum(value, @newline_styles, :spaces)

  defp normalize_value(:code_block_style, value),
    do: normalize_enum(value, @code_block_styles, :backticks)

  defp normalize_value(:preprocessing, value), do: PreprocessingOptions.new(value)

  defp normalize_value(:strong_em_symbol, value) when is_binary(value) do
    cond do
      value == "" -> "*"
      String.length(value) == 1 -> value
      true -> String.slice(value, 0, 1)
    end
  end

  defp normalize_value(:strong_em_symbol, value) when is_atom(value) do
    value |> Atom.to_string() |> normalize_value(:strong_em_symbol)
  end

  defp normalize_value(:keep_inline_images_in, value), do: normalize_set(value)
  defp normalize_value(:strip_tags, value), do: normalize_set(value)
  defp normalize_value(:preserve_tags, value), do: normalize_set(value)

  defp normalize_value(:wrap_width, value) when is_integer(value) and value > 0, do: value
  defp normalize_value(:list_indent_width, value) when is_integer(value) and value > 0, do: value
  defp normalize_value(_, value) when is_boolean(value), do: value
  defp normalize_value(_, value) when is_binary(value), do: value
  defp normalize_value(_, value) when is_integer(value), do: value
  defp normalize_value(_, value), do: value

  defp normalize_enum(value, allowed, _default) when is_atom(value) do
    if value in allowed, do: value, else: value
  end

  defp normalize_enum(value, allowed, _default) when is_binary(value) do
    normalized =
      value
      |> String.trim()
      |> String.downcase()
      |> String.replace("-", "_")

    Enum.find_value(allowed, fn atom ->
      if Atom.to_string(atom) == normalized, do: atom, else: nil
    end) || :invalid
  end

  defp normalize_enum(_value, _allowed, default), do: default

  defp normalize_set(%MapSet{} = set), do: MapSet.new(Enum.map(set, &normalize_string!/1))
  defp normalize_set(nil), do: MapSet.new()

  defp normalize_set(value) when is_list(value),
    do: MapSet.new(Enum.map(value, &normalize_string!/1))

  defp normalize_set(value) when is_binary(value), do: MapSet.new([normalize_string!(value)])
  defp normalize_set(value) when is_atom(value), do: MapSet.new([normalize_string!(value)])
  defp normalize_set(_), do: MapSet.new()

  defp normalize_string!(value) when is_binary(value), do: value
  defp normalize_string!(value) when is_atom(value), do: Atom.to_string(value)
  defp normalize_string!(value), do: to_string(value)

  @doc false
  @spec to_map(t()) :: map()
  def to_map(%__MODULE__{} = opts) do
    %{
      "heading_style" => Atom.to_string(opts.heading_style),
      "list_indent_type" => Atom.to_string(opts.list_indent_type),
      "list_indent_width" => opts.list_indent_width,
      "bullets" => opts.bullets,
      "strong_em_symbol" => opts.strong_em_symbol,
      "escape_asterisks" => opts.escape_asterisks,
      "escape_underscores" => opts.escape_underscores,
      "escape_misc" => opts.escape_misc,
      "escape_ascii" => opts.escape_ascii,
      "code_language" => opts.code_language,
      "encoding" => opts.encoding,
      "autolinks" => opts.autolinks,
      "default_title" => opts.default_title,
      "keep_inline_images_in" => mapset_to_list(opts.keep_inline_images_in),
      "br_in_tables" => opts.br_in_tables,
      "hocr_spatial_tables" => opts.hocr_spatial_tables,
      "highlight_style" => Atom.to_string(opts.highlight_style),
      "extract_metadata" => opts.extract_metadata,
      "whitespace_mode" => Atom.to_string(opts.whitespace_mode),
      "strip_newlines" => opts.strip_newlines,
      "wrap" => opts.wrap,
      "wrap_width" => opts.wrap_width,
      "strip_tags" => mapset_to_list(opts.strip_tags),
      "preserve_tags" => mapset_to_list(opts.preserve_tags),
      "convert_as_inline" => opts.convert_as_inline,
      "sub_symbol" => opts.sub_symbol,
      "sup_symbol" => opts.sup_symbol,
      "newline_style" => Atom.to_string(opts.newline_style),
      "code_block_style" => Atom.to_string(opts.code_block_style),
      "preprocessing" => PreprocessingOptions.to_map(opts.preprocessing),
      "debug" => opts.debug,
      "skip_images" => opts.skip_images
    }
  end

  defp mapset_to_list(%MapSet{} = set) do
    set
    |> Enum.map(&normalize_string!/1)
    |> Enum.reject(&(&1 == ""))
  end
end

File: packages/r/R/options.R
#' Create conversion options for html-to-markdown.
#'
#' Returns a named list of conversion options to pass to conversion functions.
#' All parameters are optional; NULL values are omitted.
#'
#' @param heading_style Style for headings: "atx", "atx_closed", or "underlined".
#' @param list_indent_type Indent type for lists: "spaces" or "tabs".
#' @param list_indent_width Number of spaces/tabs per indent level.
#' @param bullets Characters to use for bullet points.
#' @param strong_em_symbol Character for strong/emphasis: "*" or "_".
#' @param escape_asterisks Whether to escape asterisks.
#' @param escape_underscores Whether to escape underscores.
#' @param escape_misc Whether to escape miscellaneous characters.
#' @param escape_ascii Whether to escape ASCII characters.
#' @param code_language Default language for code blocks.
#' @param encoding Input encoding (e.g., "utf-8").
#' @param autolinks Whether to use autolinks for URLs.
#' @param default_title Whether to use default title attributes.
#' @param keep_inline_images_in Tags to preserve inline images in.
#' @param br_in_tables Whether to use br tags in table cells.
#' @param highlight_style Highlight style: "double_equal", "html", "bold", "none".
#' @param extract_metadata Whether to extract metadata.
#' @param whitespace_mode Whitespace handling: "normalized" or "strict".
#' @param strip_newlines Whether to strip newlines.
#' @param wrap Whether to wrap text.
#' @param wrap_width Maximum line width for wrapping.
#' @param strip_tags Tags to strip from output.
#' @param preserve_tags Tags to preserve in output.
#' @param convert_as_inline Whether to convert as inline.
#' @param sub_symbol Symbol for subscript.
#' @param sup_symbol Symbol for superscript.
#' @param newline_style Newline style: "spaces" or "backslash".
#' @param code_block_style Code block style: "indented", "backticks", or "tildes".
#' @param preprocessing Named list with preprocessing options.
#' @param debug Whether to enable debug output.
#' @return A named list of conversion options.
#' @export
conversion_options <- function(
  heading_style = NULL,
  list_indent_type = NULL,
  list_indent_width = NULL,
  bullets = NULL,
  strong_em_symbol = NULL,
  escape_asterisks = NULL,
  escape_underscores = NULL,
  escape_misc = NULL,
  escape_ascii = NULL,
  code_language = NULL,
  encoding = NULL,
  autolinks = NULL,
  default_title = NULL,
  keep_inline_images_in = NULL,
  br_in_tables = NULL,
  highlight_style = NULL,
  extract_metadata = NULL,
  whitespace_mode = NULL,
  strip_newlines = NULL,
  wrap = NULL,
  wrap_width = NULL,
  strip_tags = NULL,
  preserve_tags = NULL,
  convert_as_inline = NULL,
  sub_symbol = NULL,
  sup_symbol = NULL,
  newline_style = NULL,
  code_block_style = NULL,
  preprocessing = NULL,
  debug = NULL
) {
  opts <- list()

  if (!is.null(heading_style)) opts$heading_style <- heading_style
  if (!is.null(list_indent_type)) opts$list_indent_type <- list_indent_type
  if (!is.null(list_indent_width)) opts$list_indent_width <- as.integer(list_indent_width)
  if (!is.null(bullets)) opts$bullets <- bullets
  if (!is.null(strong_em_symbol)) opts$strong_em_symbol <- strong_em_symbol
  if (!is.null(escape_asterisks)) opts$escape_asterisks <- escape_asterisks
  if (!is.null(escape_underscores)) opts$escape_underscores <- escape_underscores
  if (!is.null(escape_misc)) opts$escape_misc <- escape_misc
  if (!is.null(escape_ascii)) opts$escape_ascii <- escape_ascii
  if (!is.null(code_language)) opts$code_language <- code_language
  if (!is.null(encoding)) opts$encoding <- encoding
  if (!is.null(autolinks)) opts$autolinks <- autolinks
  if (!is.null(default_title)) opts$default_title <- default_title
  if (!is.null(keep_inline_images_in)) opts$keep_inline_images_in <- keep_inline_images_in
  if (!is.null(br_in_tables)) opts$br_in_tables <- br_in_tables
  if (!is.null(highlight_style)) opts$highlight_style <- highlight_style
  if (!is.null(extract_metadata)) opts$extract_metadata <- extract_metadata
  if (!is.null(whitespace_mode)) opts$whitespace_mode <- whitespace_mode
  if (!is.null(strip_newlines)) opts$strip_newlines <- strip_newlines
  if (!is.null(wrap)) opts$wrap <- wrap
  if (!is.null(wrap_width)) opts$wrap_width <- as.integer(wrap_width)
  if (!is.null(strip_tags)) opts$strip_tags <- strip_tags
  if (!is.null(preserve_tags)) opts$preserve_tags <- preserve_tags
  if (!is.null(convert_as_inline)) opts$convert_as_inline <- convert_as_inline
  if (!is.null(sub_symbol)) opts$sub_symbol <- sub_symbol
  if (!is.null(sup_symbol)) opts$sup_symbol <- sup_symbol
  if (!is.null(newline_style)) opts$newline_style <- newline_style
  if (!is.null(code_block_style)) opts$code_block_style <- code_block_style
  if (!is.null(preprocessing)) opts$preprocessing <- preprocessing
  if (!is.null(debug)) opts$debug <- debug

  opts
}
